core/apple/src/AppleCore.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.NSInputStream
+import platform.darwin.UInt8Var
+
+/**
+ * Returns [RawSource] that reads from an input stream.
+ *
+ * Use [RawSource.buffered] to create a buffered source from it.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.inputStreamAsSource
+ */
+public fun NSInputStream.asSource(): RawSource = NSInputStreamSource(this)
+
+private open class NSInputStreamSource(
+    private val input: NSInputStream,
+) : RawSource {
+
+    init {
+        input.open()
```

Comment:
what will be if `NSInputStream` is already opened or closed? (before converting to source)

Reply:
`open()` should be idempotent, at least in the default `NSInputStream` implementations. There's no way to check the open/closed state of an `NSInputStream` other than getting an error if using it when closed.

Reply:
Got it!

But I have one more minor: is it ok to open stream, even if it will be not used afterwards for some reason?
May be we should open it on first read?

Im not entirely know, how it used in iOS, so my questions can be wrong 😀 
Just trying to understand which behaviour is desirable here

Reply:
Actually, I take that last part back. (It's JVM's `InputStream` you can't get the open/closed status of.) There is [`NSStreamStatus`](https://developer.apple.com/documentation/foundation/nsstreamstatus).

Reply:
I can also implement the status for `asNSInputStream()`. The question would be, should a `SourceNSInputStream` require calling `open()`, even though it's a no-op and unnecessary, or rather the status be `NSStreamStatusOpen` by default?

Mirroring the behavior of `[NSInputStream initWithData:]`, it is not open by default and reads fail until the stream is opened. So probably best to behave this way as well.